node/src/components/consensus/protocols/zug/des_testing.rs

```diff
@@ -0,0 +1,1234 @@
+            | ZugMessage::QueueAction(_)
+            | ZugMessage::FinalizedBlock(_)
+            | ZugMessage::ValidateConsensusValue(_, _)
+            | ZugMessage::NewEvidence(_)
+            | ZugMessage::Disconnect(_)
+            | ZugMessage::HandledProposedBlock(_)
+            | ZugMessage::WeAreFaulty
+            | ZugMessage::DoppelgangerDetected
+            | ZugMessage::FttExceeded
+            | ZugMessage::SendEvidence(_, _) => {
+                DeliverySchedule::AtInstant(base_delivery_timestamp + TimeDiff::from_millis(1))
+            }
+        }
+    }
+}
+
+impl ZugTestHarnessBuilder<InstantDeliveryNoDropping> {
+    fn new() -> Self {
+        ZugTestHarnessBuilder {
+            max_faulty_validators: 10,
+            faulty_percent: 0,
+            fault_type: None,
+            ftt: None,
+            consensus_values_count: 10,
+            delivery_distribution: Distribution::Uniform,
+            delivery_strategy: InstantDeliveryNoDropping,
+            weight_limits: (1, 100),
+            start_time: Timestamp::zero(),
+            end_height: TEST_END_HEIGHT,
+            weight_distribution: Distribution::Uniform,
+            config: Default::default(),
+        }
+    }
+}
+
+impl<DS: DeliveryStrategy> ZugTestHarnessBuilder<DS> {
+    /// Sets a percentage of weight that will be assigned to malicious nodes.
+    /// `faulty_weight` must be a value between 0 (inclusive) and 33 (inclusive).
+    pub(crate) fn faulty_weight_perc(mut self, faulty_weight: u64) -> Self {
+        self.faulty_percent = faulty_weight;
+        self
+    }
+
+    fn fault_type(mut self, fault_type: DesFault) -> Self {
+        self.fault_type = Some(fault_type);
+        self
+    }
+
+    pub(crate) fn consensus_values_count(mut self, count: u8) -> Self {
+        assert!(count > 0);
+        self.consensus_values_count = count;
+        self
+    }
+
+    pub(crate) fn weight_limits(mut self, lower: u64, upper: u64) -> Self {
+        assert!(
+            lower >= 100,
+            "Lower limit has to be higher than 100 to avoid rounding problems."
+        );
+        self.weight_limits = (lower, upper);
+        self
+    }
+
+    fn max_faulty_validators(mut self, max_faulty_count: u8) -> Self {
+        self.max_faulty_validators = max_faulty_count;
+        self
+    }
+
+    fn build(self, rng: &mut NodeRng) -> Result<ZugTestHarness<DS>, BuilderError> {
+        let consensus_values = (0..self.consensus_values_count)
+            .map(|el| ConsensusValue(vec![el]))
+            .collect::<VecDeque<ConsensusValue>>();
+
+        let instance_id = TEST_INSTANCE_ID;
+        let start_time = self.start_time;
+
+        let (lower, upper) = {
+            let (l, u) = self.weight_limits;
+            if l >= u {
+                return Err(BuilderError::WeightLimits);
+            }
+            (l, u)
+        };
+
+        let (faulty_weights, honest_weights): (Vec<Weight>, Vec<Weight>) = {
```

Comment:
I think this could be improved by letting a vector of validator weights and related faults just be an input to the builder, as it stands it takes some effort to understand how, e.g., the # of validators can vary here

Reply:
I'm a bit reluctant to do that, because then each test would have to randomize the validators and faults, if we wanted to keep the same level of randomness. Or we'd have to come up with an API that supports both options, but I'm not sure if that would be worth it, TBH.